### PR TITLE
bpo-45527: Don't count cache hits, just misses.

### PR DESCRIPTION
**Pull request by markshannon**
[~1% faster](https://gist.github.com/markshannon/dcee46e2086c8a9993c6fe5d0a939a33)

https://bugs.python.org/issue45527

### CODE REVIEW

**Comment by markshannon** [comment]
This will need updating once https://github.com/python/cpython/pull/29033 is merged.

**Review by Fidget-Spinner — APPROVED** [review]

File: Python/specialize.c [hunk 1]
```diff
@@ -412,6 +412,11 @@ _Py_Quicken(PyCodeObject *code) {
     return 0;
 }
 
+static inline int
+initial_counter_value(void) {
+    /* Choose a prime number ~50 */
```
Comment (Fidget-Spinner) [inline Python/specialize.c]: Wow I'm confused, how did you decide on this number?

Comment (markshannon) [reply Python/specialize.c]: It is arbitrary. Prime numbers tend to behave better than non-primes for this sort of thing, but I expect any number between ~20 and ~200 would work fine.

**Comment by markshannon** [comment]
@Fidget-Spinner did you mean to add the DO-NOT-MERGE label?

File: Python/specialize.c (hunk 1 above)
Comment (ambv) [reply Python/specialize.c]: Interesting! Why is it better than starting from 0?

Comment (ambv) [reply Python/specialize.c]: Or rather 1, right?

File: Python/ceval.c [hunk 2]
```diff
@@ -13,7 +13,7 @@
 #include "pycore_abstract.h"      // _PyIndex_Check()
 #include "pycore_call.h"          // _PyObject_FastCallDictTstate()
 #include "pycore_ceval.h"         // _PyEval_SignalAsyncExc()
-#include "pycore_code.h"          // saturating_increment()
+#include "pycore_code.h"
```
Comment (ambv) [inline Python/ceval.c:16]: To stay consistent with other includes here maybe let's do
```suggestion
#include "pycore_code.h"          // PyCodeObject_IncrementWarmup()
```

File: Python/specialize.c (hunk 1 above)
Comment (markshannon) [reply Python/specialize.c]: We want to test against zero for speed, so we want to count from N down to 0.
Counting up from -N to 0 would also work.

Comment (ambv) [inline Python/specialize.c]: Does it make sense to keep the explanation from the now-removed `saturating_start`?

```suggestion
    /* Starting value for the counter.
     * Technically this should be 1, but that is likely to
     * cause a bit of thrashing when we optimize then get an immediate miss.
     * We want to give the counter a change to stabilize, so we start at
     * a prime number close to 50.
     */
```

File: Python/ceval.c (hunk 2 above)
Comment (markshannon) [reply Python/ceval.c:16]: I don't like those comments. It suggests that `PyCodeObject_IncrementWarmup` is the only function or struct we need, but there are several.

Comment (ambv) [reply Python/ceval.c:16]: Fair enough.

File: Python/specialize.c (hunk 1 above)
Comment (ambv) [reply Python/specialize.c]: Thanks for the clarification.

**Pull request merged**